File: app/src/main/java/com/google/android/stardroid/layers/HorizonLayer.java
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.stardroid.layers;

import android.content.res.Resources;
import android.graphics.Color;

import com.google.android.stardroid.R;
import com.google.android.stardroid.base.Lists;
import com.google.android.stardroid.base.TimeConstants;
import com.google.android.stardroid.control.AstronomerModel;
import com.google.android.stardroid.renderer.RendererObjectManager.UpdateType;
import com.google.android.stardroid.source.AbstractAstronomicalSource;
import com.google.android.stardroid.source.AstronomicalSource;
import com.google.android.stardroid.source.Sources;
import com.google.android.stardroid.source.LinePrimitive;
import com.google.android.stardroid.source.TextPrimitive;
import com.google.android.stardroid.math.Vector3;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

/**
 * Creates a mark at the zenith, nadir and cardinal point and a horizon.
 *
 * @author Brent Bryan
 * @author John Taylor
 */
public class HorizonLayer extends AbstractSourceLayer {
  private final AstronomerModel model;

  public HorizonLayer(AstronomerModel model, Resources resources) {
    super(resources, true);
    this.model = model;
  }

  @Override
  protected void initializeAstroSources(ArrayList<AstronomicalSource> sources) {
    sources.add(new HorizonSource(model, getResources()));
  }

  @Override
  public int getLayerDepthOrder() {
    return 90;
  }
  
  // TODO(brent): Remove this.
  @Override
  public String getPreferenceId() {
    return "source_provider.5";
  }

  @Override
  public String getLayerName() {
    // TODO(johntaylor): i18n
    return "Horizon";
  }

  @Override
  protected int getLayerNameId() {
    return R.string.show_horizon_pref;  // TODO(johntaylor): rename this string id
  }

  /** Implementation of {@link AstronomicalSource} for the horizon source. */
  static class HorizonSource extends AbstractAstronomicalSource {
    // Due to a bug in the G1 rendering code text and lines render in different
    // colors.
    private static final int LINE_COLOR = Color.argb(120, 86, 176, 245);
    private static final int LABEL_COLOR = Color.argb(120, 245, 176, 86);
    private static final long UPDATE_FREQ_MS = 1L * TimeConstants.MILLISECONDS_PER_SECOND;

    private final Vector3 zenith = new Vector3(0, 0, 0);
    private final Vector3 nadir = new Vector3(0, 0, 0);
    private final Vector3 north = new Vector3(0, 0, 0);
    private final Vector3 south = new Vector3(0, 0, 0);
    private final Vector3 east = new Vector3(0, 0, 0);
    private final Vector3 west = new Vector3(0, 0, 0);

    private final ArrayList<LinePrimitive> linePrimitives = new ArrayList<LinePrimitive>();
    private final ArrayList<TextPrimitive> textPrimitives = new ArrayList<TextPrimitive>();
    private final AstronomerModel model;

    private long lastUpdateTimeMs = 0L;

    public HorizonSource(AstronomerModel model, Resources res) {
      this.model = model;

      List<Vector3> vertices = Lists.asList(north, east, south, west, north);
      linePrimitives.add(new LinePrimitive(LINE_COLOR, vertices, 1.5f));

      textPrimitives.add(new TextPrimitive(zenith, res.getString(R.string.zenith), LABEL_COLOR));
      textPrimitives.add(new TextPrimitive(nadir, res.getString(R.string.nadir), LABEL_COLOR));
      textPrimitives.add(new TextPrimitive(north, res.getString(R.string.north), LABEL_COLOR));
      textPrimitives.add(new TextPrimitive(south, res.getString(R.string.south), LABEL_COLOR));
      textPrimitives.add(new TextPrimitive(east, res.getString(R.string.east), LABEL_COLOR));
      textPrimitives.add(new TextPrimitive(west, res.getString(R.string.west), LABEL_COLOR));
    }

    private void updateCoords() {
      // Blog.d(this, "Updating Coords: " + (model.getTime().getTime() - lastUpdateTimeMs));

      this.lastUpdateTimeMs = model.getTime().getTime();
      this.zenith.assign(model.getZenith());
      this.nadir.assign(model.getNadir());
      this.north.assign(model.getNorth());
      this.south.assign(model.getSouth());
      this.east.assign(model.getEast());
      this.west.assign(model.getWest());
    }

    @Override
    public Sources initialize() {
      updateCoords();
      return this;
    }

    @Override
    public EnumSet<UpdateType> update() {
      EnumSet<UpdateType> updateTypes = EnumSet.noneOf(UpdateType.class);

      // TODO(brent): Add distance here.
      if (Math.abs(model.getTime().getTime() - lastUpdateTimeMs) > UPDATE_FREQ_MS) {
        updateCoords();
        updateTypes.add(UpdateType.UpdatePositions);
      }
      return updateTypes;
    }

    @Override
    public List<TextPrimitive> getLabels() {
      return textPrimitives;
    }

    @Override
    public List<LinePrimitive> getLines() {
      return linePrimitives;
    }
  }
}

File: app/src/main/java/com/google/android/stardroid/layers/HorizonLayer.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid.layers

import android.content.SharedPreferences
import android.content.res.Resources
import com.google.android.stardroid.R
import com.google.android.stardroid.base.TimeConstants
import com.google.android.stardroid.control.AstronomerModel
import com.google.android.stardroid.math.Vector3
import com.google.android.stardroid.renderables.*
import com.google.android.stardroid.renderer.RendererObjectManager.UpdateType
import java.util.*
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.sin

/**
 * Creates a mark at the zenith, nadir and cardinal point and a horizon.
 *
 * @author Brent Bryan
 * @author John Taylor
 */
class HorizonLayer(private val model: AstronomerModel, resources: Resources, preferences: SharedPreferences) :
    AbstractRenderablesLayer(resources, true, preferences) {
    override fun initializeAstroSources(sources: ArrayList<AstronomicalRenderable>) {
        sources.add(HorizonRenderable(model, resources))
    }

    override val layerDepthOrder = 90

    // TODO(brent): Remove this.
    override val preferenceId = "source_provider.5"

    // TODO(johntaylor): i18n
    override val layerName = "Horizon"

    override val layerNameId = R.string.show_horizon_pref // TODO(johntaylor): rename this string id

    /** Implementation of [AstronomicalRenderable] for the horizon source.  */
    internal class HorizonRenderable(private val model: AstronomerModel, resources: Resources) :
        AbstractAstronomicalRenderable() {
        private val zenith = Vector3(0f, 0f, 0f)
        private val nadir = Vector3(0f, 0f, 0f)
        private val north = Vector3(0f, 0f, 0f)
        private val south = Vector3(0f, 0f, 0f)
        private val east = Vector3(0f, 0f, 0f)
        private val west = Vector3(0f, 0f, 0f)
        // NUM_SEGMENTS+1 because the last vertex closes the loop back to the first.
        private val horizonVerts: Array<Vector3> = Array(NUM_SEGMENTS + 1) { Vector3(0f, 0f, 0f) }
        // 25 narrow glow rings, each recomputed from horizonVerts in updateCoords().
        private val glowRings: Array<Array<Vector3>> =
            Array(NUM_GLOW_RINGS) { Array(NUM_SEGMENTS + 1) { Vector3(0f, 0f, 0f) } }
        private var lastUpdateTimeMs = 0L

        private fun updateCoords() {
            lastUpdateTimeMs = model.time.time
            zenith.assign(model.zenith)
            nadir.assign(model.nadir)
            north.assign(model.north)
            south.assign(model.south)
            east.assign(model.east)
            west.assign(model.west)

            // Pre-compute tilt sin/cos for each glow ring once per update.
            val cosTilts = FloatArray(NUM_GLOW_RINGS)
            val sinTilts = FloatArray(NUM_GLOW_RINGS)
            for (ringIdx in 0 until NUM_GLOW_RINGS) {
                val tiltRad = Math.toRadians(((ringIdx + 1) * GLOW_RING_SPACING_DEG).toDouble())
                cosTilts[ringIdx] = cos(tiltRad).toFloat()
                sinTilts[ringIdx] = sin(tiltRad).toFloat()
            }

            // Horizon circle: p(θ) = north·cos(θ) + east·sin(θ)
            for (i in 0..NUM_SEGMENTS) {
                val angle = 2.0 * Math.PI * i / NUM_SEGMENTS
                val cosA = cos(angle).toFloat()
                val sinA = sin(angle).toFloat()
                val hx = north.x * cosA + east.x * sinA
                val hy = north.y * cosA + east.y * sinA
                val hz = north.z * cosA + east.z * sinA
                horizonVerts[i].assign(hx, hy, hz)
                // Tilt each glow ring toward nadir: p_tilted = p·cos(t) + nadir·sin(t)
                for (ringIdx in 0 until NUM_GLOW_RINGS) {
                    glowRings[ringIdx][i].assign(
                        hx * cosTilts[ringIdx] + nadir.x * sinTilts[ringIdx],
                        hy * cosTilts[ringIdx] + nadir.y * sinTilts[ringIdx],
                        hz * cosTilts[ringIdx] + nadir.z * sinTilts[ringIdx]
                    )
                }
            }
        }

        override fun initialize(): Renderable {
            updateCoords()
            return this
        }

        override fun update(): EnumSet<UpdateType> {
            val updateTypes = EnumSet.noneOf(UpdateType::class.java)
            if (abs(model.time.time - lastUpdateTimeMs) > UPDATE_FREQ_MS) {
                updateCoords()
                updateTypes.add(UpdateType.UpdatePositions)
            }
            return updateTypes
        }

        override val labels: MutableList<TextPrimitive> = ArrayList()
        override val lines: MutableList<LinePrimitive> = ArrayList()

        companion object {
            private const val UPDATE_FREQ_MS = 1L * TimeConstants.MILLISECONDS_PER_SECOND
            // 180 segments per ring: visually smooth, and keeps total quad count
            // (26 rings × 180) × 4 vertices = 18 720 — safely under the signed-short
            // index limit of 32 767 in PolyLineObjectManager.
            private const val NUM_SEGMENTS = 180
            // 25 narrow rings spaced 0.5° apart → smooth exponential glow from horizon
            // down to ~12.5°. lineWidth 14 ≈ 0.97° half-width, so adjacent rings overlap
            // by ~1.4° and each interior point is covered by ~3 rings simultaneously,
            // making the per-ring alpha steps imperceptible.
            private const val NUM_GLOW_RINGS = 25
            private const val GLOW_RING_SPACING_DEG = 0.5f
            private const val GLOW_LINE_WIDTH = 14f
            private const val GLOW_ALPHA_BASE = 0.50f
            private const val GLOW_ALPHA_DECAY = 0.20f  // per ring index (natural units)
        }

        init {
            val lineColor = resources.getColor(R.color.horizon_line, null)
            val labelColor = resources.getColor(R.color.horizon_label, null)
            lines.add(LinePrimitive(lineColor, horizonVerts.toList(), 2.5f))

            val baseAlpha = android.graphics.Color.alpha(lineColor)
            val r = android.graphics.Color.red(lineColor)
            val g = android.graphics.Color.green(lineColor)
            val b = android.graphics.Color.blue(lineColor)
            for (i in 0 until NUM_GLOW_RINGS) {
                val alpha = (GLOW_ALPHA_BASE * exp(-i * GLOW_ALPHA_DECAY) * baseAlpha)
                    .toInt().coerceIn(0, 255)
                val glowColor = android.graphics.Color.argb(alpha, r, g, b)
                lines.add(LinePrimitive(glowColor, glowRings[i].toList(), GLOW_LINE_WIDTH))
            }
            labels.add(TextPrimitive(zenith, resources.getString(R.string.zenith), labelColor))
            labels.add(TextPrimitive(nadir, resources.getString(R.string.nadir), labelColor))
            labels.add(TextPrimitive(north, resources.getString(R.string.north), labelColor))
            labels.add(TextPrimitive(south, resources.getString(R.string.south), labelColor))
            labels.add(TextPrimitive(east, resources.getString(R.string.east), labelColor))
            labels.add(TextPrimitive(west, resources.getString(R.string.west), labelColor))
        }
    }
}
